Implement the /cash API

jniles opened the pull request:
This PR implements the `/cash` API and provides integration tests to document and maintain functionality.  Both caution and patient invoice payments are supported by providing the property `is_caution` in the POST request.

_Supported Routes_
- `GET /cash`  : retrieve a list of all cash payments
- `GET /cash/:uuid` : retrieve the details of a single cash payment
- `POST /cash` : create a cash payment (either caution or payment against invoices)
- `PUT /cash/:uuid` : update non-protected (non-financial) fields on a previous cash payment

Note that `DELETE` methods have not been implemented.

Additional changes to allow seamless testing:
1. I've added a file called `helpers.js` in the integration testing directory that performs routine actions such as logging in, setting the appropriate TLS flags, adding promises in case of low NodeJS versions, etc.
2. The cash table has dropped many columns.  These changes are recorded in `update/schema-v2.sql` which should provide a migration from `bhima-1.X` -> `bhima-2.X`.
3. Both `sh/rebuild-db.sh` and `sh/integration-tests.sh` have been updated to build the database from `schema.sql`, `data.sql`, and `update/schema-v2.sql`.
4. The database will always default to UTF-8 for multi-lingual support.
5. Cash record creation uses triggers introduced by @sfount in #9.
6. A small script, `migrate-previous-dump.sh` was created to upgrade the data from a live `bhima-1.X` database to `bhima-2.X` to ease data migrations in the future.

Note: for full usage of the new shell scripts, see the associated commits.  They should provide sufficient descriptions to use them.

**EDIT** The functionality provided here should be enough to build the Caisse Aux module.  A separate PR will provide the functionality needed for the Cash Discard/Debit Note module.

TODO
- [ ] Complete the posting journal integration to allow writing to the journal.

jniles committed:
CHORE(migration): added database migration scripts

This commit adds important database migration scripts.  This includes a
shell script for importing data from a bhima-1.X dump file and schema
upgrades to be run on a live bhima-1.X database.

Case 1: To upgrade from a dumped file (such as the dropbox backup), the
correct approach is to unzip the file and run the script
./sh/migrate-previous-dump.sh on the file.  This script will"
  a) Change all CHARSETs to UTF-8
  b) Change all ENGINEs to InnoDB
  c) Reset all AUTO_INCREMENTs to 0

Case 2: To migrate the bhima-1.X schema to the bhima-2.X schema, use the
SQL files in update/schema-v2.sql   This SQL file uses temporary tables
to ensure that all data is properly copied over to a new schema.

jniles committed:
TEST(cash): added /cash integration tests

This commit adds ~12 failing /cash API endpoint tests for the caution/invoice
payment functionality of the cash module.

Additionally, a test helper file (server/test/api/helper.js) is added to
handle routine test parameters without removing readability.  The /cash
API is the first to leverage this helpers and can be used as a reference
while the work is expanded to other files.

jniles committed:
CHORE(test): update tests to use test helpers

This commit updates most existing tests to use test helpers for better
readability and maintainability.  There is also a minor bug fix due to a
typo in the patient API tests.

jniles committed:
CHORE(cash): GET API uses new db schema

This commit finalizes the GET routes in the cash API.  Both GET /cash
and /cash/:uuid have been implemented and pass integration tests.

Additionally, the build-db.sh and integration-test.sh scripts now build
the schema-v2.sql file last.  This lets us test the newer versions of
the database easily.

jniles committed:
TEST(cash): create caution tests pass

This commit fixes the POST route for /cash to allow caution payments.
It includes a passing integration test as proof of work.

jniles committed:
TEST(cash): finished POST integration tests

This commit finalizes the POST integration testing for the cash payments
interface.  Supported routes are:
  - GET /cash
  - GET /cash/:uuid
  - POST /cash (with is_caution = 1 to record a caution)
  - POST /cash (with is_caution = 0 to record a sale payment)
  - PUT /cash/:uuid (updates non-financial fields)

There is a design decision whether DELETE /cash/:uuid should create a
debit note or simply delete the records.  Debit note tests are written,
but currently "pending" using `describe.skip()`.

jniles committed:
CHORE(cash): improve test readability

DedrickEnc commented on server/config/routes.js:
```diff
@@ -360,6 +361,13 @@ exports.configure = function (app) {
   app.put('/cashboxes/:id/currencies/:currencyId', cashboxes.currencies.update);
 
+  // cash (aux/primary)
+  app.get('/cash', cash.list);
```
The route must be in plural `/cashes` instead of `/cash`

jniles replied:
Cash is plural in English. :).

DedrickEnc commented on server/config/routes.js:
```diff
@@ -360,6 +361,13 @@ exports.configure = function (app) {
+  app.get('/cash/:uuid', cash.details);
+  app.post('/cash', cash.create);
+  app.put('/cash/:uuid', cash.update);
+  app.delete('/cash/:uuid', cash.debitNote);
```
In you PR description you said there is no delete route for cash, but why you keep it here?

jniles replied:
I kept it in as a placeholder for a future "cash_discard" route.  It doesn't actually do anything yet.  If we decide to not use the DELETE request for cash_discard, I'll remove it and make a separate set of routes like `cash/discard/:uuid`.

DedrickEnc commented on server/controllers/finance/cash.js:
```diff
@@ -0,0 +1,247 @@
+
+// check if an object is empty
+// TODO - can this be from a library?
+function empty(object) {
```
The server side API should not do validation and checking, that should be handled at the server side

jniles replied:
What do you mean?

DedrickEnc replied:
@jniles 
The server API is essentially a data access layer, checking and validation must be done at a high level.

jniles replied:
So, where should this line of code go?

DedrickEnc replied:
Client side service, form validation...

jniles replied:
It is unacceptable to perform all validation on the client.  Client-side validation is mainly for UI/UX, **NOT** data integrity.  The server is the only reliable line of defense against corrupt data sent from either a) a poorly designed client service or b) an active attacker.

DedrickEnc replied:
I do not say, We must no have a if instruction at the server side. The Data Access Layer is not responsible of validation, the database him self is responsible of data integrity witch is the ACID properties  (Atomicity, Coherence, Isolation, Durability).

The DAL or DAO must be independent from business layer, doing validation in DAL is to encrust some business logic in it so the DAO will be dependent to the Business logic layer, the problem is when the enterprise will decide to change some business rule, we will be supposed to touch the Business Layer (which is essentially our client side) and The DAL, that situation is not good. Isolate DAL from BL (Business Layer) is very important.

If possible we can make our database robust by making a good implementation it is one of our goal I think.

@jniles @sfount

jniles replied:
So...  my question still stands.  Where should this line of code go, in your opinion?  Where should the business layer be, if this is the data access layer?

DedrickEnc commented on server/controllers/finance/cash.js:
```diff
@@ -0,0 +1,247 @@
+ *   ...
+ *  }
+ */
+exports.details = function details(req, res, next) {
```
Interfaces must be named as an action `getDetails()` seems meaningful than `details`

jniles replied:
Since this function will only ever be called from an HTTP GET request, the "get" in "getDetails" is implied.  Here's what it looks like from the server:

``` js
app.get('/cash/:uuid', cash.getDetails);
// versus
app.get('/cash/:uuid', cash.details);
```

In the first call, get is repeated twice, in the second one it reads cleanly.

DedrickEnc replied:
I think the first get is the http action, the second one is the API method which is exposed, an interface must be action not a simple name.
I think `list` `create` `update` `debitNote` are well named than `details` there are informing what the cash API can do (**do like action**).

jniles replied:
Okay I will change it.  Remember that this method will _only_ ever be called from an HTTP GET method.

DedrickEnc commented on server/controllers/finance/cash.js:
```diff
@@ -0,0 +1,247 @@
+  // generate a UUID if it not provided.
+  // @TODO - should we validate that this is an _actual_ uuid, or should this
+  // be sufficient?
+  data.uuid = data.uuid || uuid();
```
We should decide, which part is responsible to assign ID to the entity during creation the server or the client. not both of them.

jniles replied:
It depends.  If we ever plan to support offline (client-side) creation of cash payments, sales, patients, etc, we must provide the option for the client to generate their own IDs.  If the client must have an active connection to the server at all times, we can limit ID creation to the server.

Ultimately, the database doesn't care - a UUID will be unique no matter where it is created.  This code just allows us to avoid that making the offline/online decision until we've gotten further into development.

DedrickEnc replied:
I agree

DedrickEnc commented on server/controllers/finance/cash.js:
```diff
@@ -0,0 +1,247 @@
+
+  // loop through update keys and ensure that we are only updating non-protected
+  // fields
+  var keys = Object.keys(req.body);
```
This test can be performed at the high level

jniles replied:
[According the Open Web Application Security Project](https://www.owasp.org/index.php/REST_Security_Cheat_Sheet),

> A data contextual check needs to be done, server side, with each request.

From an [HTML5 Rocks article](http://www.html5rocks.com/en/tutorials/forms/constraintvalidation/) on client-side validation:

> It's important to remember that even with these new APIs client side validation does not remove the need for server side validation. Malicious users can easily workaround any client side constraints, and, HTTP requests don't have to originate from a browser.
> 
> Therefore client side validation should always be treated as a progressive enhancement to the user experience; all forms should be usable even if client side validation is not present.
> 
> Since you have to have server side validation anyways, if you simply have your server side code return reasonable error messages and display them to the end user you have a built in fallback for browsers that don't support any form of client side validation.

DedrickEnc commented on server/controllers/finance/cash.js:
```diff
@@ -0,0 +1,247 @@
+  .then(function (record) {
+
+    // if we get here, we know we have a cash record by this UUID.
+    // we can try to update it.
```
Where do you use te close where for your update request?

jniles replied:
~~Aha!  Good catch.~~  Fixed in latest commit.

jniles committed:
FEEDBACK(cash): renamed GET route, syntax error

This commit fixes a critical API error which omitted the ID condition on
a PUT request.  It also encorporates feedback from @dedrickenc on route
names.

DedrickEnc commented on server/models/update/schema-v2.sql:
```diff
@@ -0,0 +1,125 @@
+--
```
Sorry, Why this new file is important?

jniles replied:
The schema is very important.  The file, not so much.   `synt.sql` looked like it mostly contained data changes, so I made an upgrade SQL for the schema.  That way, we'll know what schema changes when introduced the when we try to do a 1.X --> 2.X upgrade.

This file already contains the upgrade script to migrate data in the 1.X cash table (the one found at our test hospital) to the 2.X cash table.  All we need to do is run it on the production database and everything will automatically be imported to the 2.X schema.

sfount commented on server/test/api/exchange.js:
```diff
@@ -5,49 +5,34 @@ var chaiHttp = require('chai-http');
-// base URL
-var url = 'https://localhost:8080';
-var user = { username : 'superuser', password : 'superuser', project: 1};
+var helpers = require('./helpers');
```
This is a good idea, much cleaner :+1: 

As soon as this can be pulled this should be encouraged for integration tests going forward. It can also also be applied to tests written in parallel with this pull request. 
@IMA-WorldHealth/local-contributors

jniles committed:
CHORE(merge): merge fixes with /sales API

This commit fixes data conflicts with the sales API, and updates the sales
tests to reflect the creation of sales within the test SQL.

jniles committed:
FEEDBACK(cash): rm schema-v2 in favor of synt.sql

This commit moves the work in schema-v2 and puts it in synt.sql.

jniles committed:
CHORE(merge): rm merge regressions

This commit removes changes to protractor.conf.js introduced by the PR

jniles committed:
MERGE(cash): fix conflicts with master branch
